packages/gatsby-script/src/gatsby-script.tsx

```diff
@@ -0,0 +1,189 @@
+import React, { useEffect, useContext } from "react"
+import { PartytownContext } from "./partytown-context"
+import type { ReactElement, ScriptHTMLAttributes } from "react"
+import { requestIdleCallback } from "./request-idle-callback-shim"
+
+export enum ScriptStrategy {
+  postHydrate = `post-hydrate`,
+  idle = `idle`,
+  offMainThread = `off-main-thread`,
+}
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface ScriptProps
+  extends Omit<ScriptHTMLAttributes<HTMLScriptElement>, `onLoad` | `onError`> {
+  id?: string
+  strategy?: ScriptStrategy | `post-hydrate` | `idle` | `off-main-thread`
+  children?: string
+  onLoad?: (event: Event) => void
+  onError?: (event: ErrorEvent) => void
+  forward?: Array<string>
+}
+
+const handledProps = new Set([
+  `src`,
+  `strategy`,
+  `dangerouslySetInnerHTML`,
+  `children`,
+  `onLoad`,
+  `onError`,
+])
+
+export const scriptCache = new Set()
+
+export function Script(props: ScriptProps): ReactElement | null {
+  const {
+    id,
+    src,
+    strategy = ScriptStrategy.postHydrate,
+    onLoad,
+    onError,
+  } = props || {}
+  const { collectScript } = useContext(PartytownContext)
+
+  useEffect(() => {
+    let script: HTMLScriptElement | null
+
+    switch (strategy) {
+      case ScriptStrategy.postHydrate:
+        script = injectScript(props)
+        break
+      case ScriptStrategy.idle:
+        requestIdleCallback(() => {
+          script = injectScript(props)
+        })
+        break
+      case ScriptStrategy.offMainThread:
+        if (typeof window !== `undefined` && collectScript) {
```

Comment:
useEffects only run on the client so window is always defined.
```suggestion
        if (collectScript) {
```